internal/controller/projectdevelopmentstream_controller.go

```diff
@@ -82,60 +86,97 @@ func (r *ProjectDevelopmentStreamReconciler) Reconcile(ctx context.Context, req
 		}
 	}
 
-	log.Info("Applying resources from ProjectDevelopmentStream")
+	var templateName string
+	if pds.Spec.Template == nil {
+		log.Info("No template is associated with this ProjectDevelopmentStream")
+		return ctrl.Result{}, nil
+	}
+	templateName = pds.Spec.Template.Name
+	log = log.WithValues("PDS Template", templateName)
+
+	var pdst projctlv1beta1.ProjectDevelopmentStreamTemplate
+	templateKey := client.ObjectKey{Namespace: pds.GetNamespace(), Name: templateName}
+	if err := r.Get(ctx, templateKey, &pdst); err != nil {
+		log.Error(err, "Failed to fetch template")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	log.Info("Applying resources from ProjectDevelopmentStreamTemplate")
+	resources, err := template.MkResources(pds, pdst)
+	if err != nil {
+		log.Error(err, "Failed to generate resources from template")
+		// We return 'nil' error because there is not point retrying the
+		// reconcile loop
+		return ctrl.Result{}, nil
+	}
 
-	ctrlResult := ctrl.Result{}
-	for _, resourceTemplate := range pds.Spec.Resources {
-		resource := resourceTemplate.DeepCopy()
+	var requeue bool
+	for _, resource := range resources {
 		log := log.WithValues(
 			"apiVersion", resource.GetAPIVersion(),
 			"kind", resource.GetKind(),
 			"name", resource.GetName(),
 		)
 		log.Info("Creating/Updating resource")
-		if resource.GetNamespace() != "" && resource.GetNamespace() != pds.GetNamespace() {
-			log.Info(
-				"Resource namespace set to ProjectDevelopmentStream namespace",
-				"PDS namespace", pds.GetNamespace(),
-				"resource original namespace", resource.GetNamespace(),
-			)
+		ownership.AddMissingUIDs(ctx, r.Client, resource)
+		if len(resource.GetOwnerReferences()) <= 0 {
+			// If the resource does not have an owner set, use the PDS
+			_ = controllerutil.SetOwnerReference(&pds, resource, r.Scheme)
```

Comment:
Can this resource have multiple pds depending on it?
If I understand correctly, with the logic here we allow only one owner - the first to get to this point in the code.

Suppose others come after, and at some point the first pds is being removed. Would the resource be removed (for empty owner ref), while the newer pds still depend on it?

Reply:
an Application can only belong to a single PDS, and the point of setting the ownership links is to make cascading deletes happen.

Reply:
> the point of setting the ownership links is to make cascading deletes happen.

Understood that. Wanted to make sure there cannot be other owners. Cool.